tfjs-backend-webgpu/src/kernels/ToPixels.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use backend file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {KernelConfig, KernelFunc} from '@tensorflow/tfjs-core';
+import {ToPixels, ToPixelsInputs, ToPixelsOutput} from '@tensorflow/tfjs-core';
+import {TensorInfo} from '@tensorflow/tfjs-core';
+
+import {WebGPUBackend} from '../backend_webgpu';
+// import {ToPixelsProgram} from './toPixels_webgpu';
+import {ToCanvasProgram} from './to_canvas_webgpu';
+
+export const toPixelsConfig: KernelConfig = {
+  kernelName: ToPixels,
+  backendName: 'webgpu',
+  kernelFunc: toPixels as {} as KernelFunc,
+};
+
+export function toPixels(args: {
+  inputs: ToPixelsInputs,
+  backend: WebGPUBackend,
+  output: ToPixelsOutput
+}): TensorInfo {
+  const {inputs, backend /*, output*/} = args;
+  let {$img} = inputs;
+  // const {canvas} = output;
+  const [height, width] = $img.shape.slice(0, 2);
+
+  const outShape = [height, width, 4];
+  const program = new ToCanvasProgram(outShape, $img.dtype);
```

Comment:
should the caller provide the canvas,given that some of the platform do not have the document variable.

Reply:
If the caller provide the canvas, it may need one or two extra copying and we also need to check different canvas context.
For example, if we directly return the webgpu context canvas, assume that the user want to use it in his webgl program. He can just call `texImage2D` with the webgpu context canvas as the source to get a `GLTexture` and continue to use the texture in his program.
But if we let user pass a canvas, if the canvas is a webgl canvas, we need to 1) use `texImage2D` to get a `GLTexture` from the webgpu context canvas, then 2) draw the `GLTexture` to the canvas that the user provided. 3) then the user can use this canvas, he has to call `texImage2D` with the canvas he provided again to get the `GLTexture`.  
If the canvas is a 2d canvas, we need to draw the webgpu context canvas to the 2d canvas.

Reply:
I think canvas can serve both webgl and 2d, only the context is fixed to one or the other?

Reply:
@pyu10055 Are you suggesting the new kernel API as
1) function [async] kernelName(img: Tensor2D|Tensor3D|TensorLike): HTMLCanvasElement|OffscreenCanvas

or

2) function [async ]kernelName(img: Tensor2D|Tensor3D|TensorLike, canvas: HTMLCanvasElement|OffscreenCanvas)

For 1, we can return different canvas context based on backend (I prefer this one). For 2, once the canvas the user passed has been specified to one context, we can't change it. So that's way I said there will be extra copying needed for this case.

>should the caller provide the canvas,given that some of the platform do not have the document variable.

Can `OffscreenCanvas` resolve it?

>I think canvas can serve both webgl and 2d, only the context is fixed to one or the other?

Yes. But once it's fixed to one, we can't change it to another. And we need to do different processing for them.

Reply:
I see, I was thinking option 2, do we need a 2d or webgl canvas?
can we ask users to provide the exact type of canvas that we need?

Reply:
I think for option 2, the best way is let user pass a non-context specified canvas and let backends decide what kind of context they want. In this case, there is no difference for webgpu backend since in both situation it needs an extra canvas as the draw target. But for webgl backend, it needs extra copy to move the data from the default webgl context canvas to the new passed canvas.
But if we allow user to provide a specific context canvas, the extra copy is needed both for webgl and webgpu unless we also expose the GPU device id to users. I mean for example when users register webgpu backend, we allow users specify a device id to webgpu backend, then backend can directly use this device id and don't need to create a new one. If the passed canvas context is in the same device id with the webgpu backend, no extra copying is needed.